plugins/woocommerce-admin/client/marketplace/components/category-selector/category-selector.tsx

```diff
@@ -72,21 +62,125 @@ export default function CategorySelector(
-
-				setVisibleItems( visibleCategoryItems );
-				setDropdownItems( dropdownCategoryItems );
+				setCategoriesToShow( categories );
 			} )
 			.catch( () => {
-				setVisibleItems( [] );
-				setDropdownItems( [] );
+				setCategoriesToShow( [] );
 			} )
 			.finally( () => {
 				setIsLoading( false );
 			} );
-	}, [ props.type ] );
+	}, [ props.type, setCategoriesToShow ] );
+
+	useEffect( () => {
+		// If no category is selected, show All as selected
+		let categoryToSearch = ALL_CATEGORIES_SLUGS[ props.type ];
+
+		if ( query.category ) {
+			categoryToSearch = query.category;
+		}
+
+		const selectedCategory = categoriesToShow.find(
+			( category ) => category.slug === categoryToSearch
+		);
+
+		if ( selectedCategory ) {
+			setSelected( selectedCategory );
+		}
+	}, [ query.category, props.type, categoriesToShow ] );
+
+	useEffect( () => {
+		if ( selectedCategoryRef.current ) {
+			selectedCategoryRef.current.scrollIntoView( {
+				block: 'nearest',
+				inline: 'center',
+			} );
+		}
+	}, [ selected ] );
+
+	function checkOverflow() {
+		if (
+			categorySelectorRef.current &&
+			categorySelectorRef.current.parentElement?.scrollWidth
+		) {
+			const isContentOverflowing =
+				categorySelectorRef.current.scrollWidth >
+				categorySelectorRef.current.parentElement.scrollWidth;
+
+			setIsOverflowing( isContentOverflowing );
+		}
+	}
+
+	function checkScrollPosition() {
+		const ulElement = categorySelectorRef.current;
+
+		if ( ! ulElement ) {
+			return;
+		}
+
+		const { scrollLeft, scrollWidth, clientWidth } = ulElement;
+
+		if ( scrollLeft < 10 ) {
+			setScrollPosition( 'start' );
+
+			return;
+		}
+
+		if ( scrollLeft + clientWidth < scrollWidth ) {
+			setScrollPosition( 'middle' );
+
+			return;
+		}
+
+		if ( scrollLeft + clientWidth === scrollWidth ) {
+			setScrollPosition( 'end' );
+		}
+	}
+
+	const debouncedCheckOverflow = useDebounce( checkOverflow, 300 );
+	const debouncedScrollPosition = useDebounce( checkScrollPosition, 100 );
+
+	function scrollCategories( scrollAmount: number ) {
+		if ( categorySelectorRef.current ) {
+			categorySelectorRef.current.scrollTo( {
+				left: categorySelectorRef.current.scrollLeft + scrollAmount,
+				behavior: 'smooth',
+			} );
+		}
+	}
+
+	function scrollToNextCategories() {
+		scrollCategories( 200 );
+	}
+
+	function scrollToPrevCategories() {
+		scrollCategories( -200 );
+	}
+
+	useEffect( () => {
+		window.addEventListener( 'resize', debouncedCheckOverflow );
+
+		const ulElement = categorySelectorRef.current;
+
+		if ( ulElement ) {
+			ulElement.addEventListener( 'scroll', debouncedScrollPosition );
+		}
+
+		return () => {
+			window.removeEventListener( 'resize', debouncedCheckOverflow );
+
+			if ( ulElement ) {
+				ulElement.removeEventListener(
+					'scroll',
+					debouncedScrollPosition
+				);
+			}
+		};
+	}, [ debouncedCheckOverflow, debouncedScrollPosition ] );
+
+	useEffect( () => {
+		checkOverflow();
+	}, [ categoriesToShow ] );
```

Comment:
No change needed, but it might also work to call checkOverflow() in the first useEffect hook, after we call setCategoriesToShow(). Means we don't need another, separate useEffect hook here.

Reply:
Makes sense :)